File: ShootR/ShootR/Server/Abilities/Abstractions/Ability.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;

namespace ShootR
{
    public abstract class Ability
    {
        public Ability()
        {
            Active = false;            
        }

        public bool Active { get; set; }
        public DateTime? ActivatedAt { get; private set; }

        public virtual void Activate()
        {
            ActivatedAt = DateTime.UtcNow;
            Active = true;
        }

        public virtual void Deactivate()
        {
            ActivatedAt = null;
            Active = false;
        }

        public virtual void Update(DateTime utcNow)
        {
        }
    }
}

File: ShootR/ShootR/Client/Ships/Abilities/Ability.ts
/// <reference path="../../../Scripts/endgate-0.2.0.d.ts" />

module ShootR {

    export class Ability {
        public Active: boolean;
        public ActivatedAt: Date;

        constructor(public Name: string) {
            this.Active = false;
            this.ActivatedAt = null;
        }

        public Activate(): void {
            this.Active = true;
            this.ActivatedAt = new Date();
        }

        public Deactivate(): void {
            this.Active = false;
            this.ActivatedAt = null;
        }

        // Meant to be overridden
        public Update(gameTime: eg.GameTime): void { }
    }

}